src/plugins/direct_access_links/public/direct_access_link_generator.ts

```diff
@@ -0,0 +1,174 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { DirectAccessLinksStart } from './plugin';
+
+export type GeneratorId = string;
+
+export interface GeneratorState<
+  S extends {},
+  I extends string | undefined = undefined,
+  MS extends {} | undefined = undefined
+> {
+  State: S;
+  MigratedId?: I;
+  MigratedState?: MS;
+}
+
+export interface GeneratorStateMapping {
+  // The `any` here is quite unfortunate.  Using `object` actually gives no type errors in my IDE
+  // but running `node scripts/type_check` will cause an error:
+  // examples/access_links_examples/public/direct_access_link_generator.ts:77:66 -
+  // error TS2339: Property 'name' does not exist on type 'object'.  However it's correctly
+  // typed when I edit that file.
+  [key: string]: GeneratorState<any, string | undefined, object | undefined>;
+}
+
+export interface AccessLinkGenerator<Id extends GeneratorId> {
+  id: Id;
+  createUrl(state: GeneratorStateMapping[Id]['State']): Promise<string>;
+  isDeprecated: boolean;
+  migrate(
+    state: GeneratorStateMapping[Id]['State']
+  ): Promise<{
+    state: GeneratorStateMapping[Id]['MigratedState'];
+    id: GeneratorStateMapping[Id]['MigratedId'];
+  }>;
+}
+
+const noMigrationFnProvidedWarningText = i18n.translate(
+  'directAccessLinks.error.noMigrationFnProvided',
+  {
+    defaultMessage:
+      'If the access link generator is marked as deprecated, you must provide a migration function.',
+  }
+);
+
+export interface DirectAccessLinkOptions<Id extends GeneratorId> {
+  id: Id;
+  createUrl?: (state: GeneratorStateMapping[Id]['State']) => Promise<string>;
+  isDeprecated?: boolean;
+  migrate?: (
+    state: GeneratorStateMapping[Id]['State']
+  ) => Promise<{
+    state: GeneratorStateMapping[Id]['MigratedState'];
+    id: GeneratorStateMapping[Id]['MigratedId'];
+  }>;
+}
+
+export const createDirectAccessLinkGenerator = <Id extends GeneratorId>(
+  options: DirectAccessLinkOptions<Id>,
+  getGenerator: DirectAccessLinksStart['getAccessLinkGenerator']
+): AccessLinkGenerator<Id> => {
+  if (options.isDeprecated && !options.migrate) {
+    throw new Error(noMigrationFnProvidedWarningText);
+  }
+
+  if (!options.isDeprecated && options.migrate) {
+    throw new Error(
+      i18n.translate('directAccessLinks.error.migrationFnGivenNotDeprecated', {
```

Comment:
I haven't see we translate the error messages in services, for example, in Expressions plugin. Is this something we should start doing or should the translation be up to the UI and the service should emit some language agnostic code, like

```ts
new Error('migrationFnGivenNotDeprecated');
```

or 

```ts
new Error('directAccessLinks.error.migrationFnGivenNotDeprecated');
```

or

```ts
const error = new Error('Some human readable error in English.');
error.code = 'directAccessLinks.error.migrationFnGivenNotDeprecated';
```

like [in Node](https://nodejs.org/api/errors.html#errors_error_code).